lib/jackalope/persistent_work_list.ex

```diff
@@ -0,0 +1,357 @@
+defmodule Jackalope.PersistentWorkList do
+  @moduledoc """
+  A genserver wrapper for CubQ which we leverage to store and restore worklist tasks during disconnections
+  """
+  use GenServer
+
+  alias Jackalope.WorkList.Expiration
+
+  require Logger
+
+  @default_max_size 100
+
+  @tick_delay 10 * 60 * 1_000
+
+  defmodule State do
+    @moduledoc false
+    defstruct db: nil,
+              queue: nil,
+              queue_name: nil,
+              max_work_list_size: nil,
+              expiration_fn: nil,
+              update_expiration_fn: nil
+  end
+
+  @doc "Create a new work list"
+  @spec new(function(), function(), non_neg_integer(), Keyword.t()) :: pid()
+  def new(expiration_fn, update_expiration_fn, max_size \\ @default_max_size, opts \\ []) do
+    args = [
+      expiration_fn: expiration_fn,
+      update_expiration_fn: update_expiration_fn,
+      max_size: max_size,
+      opts: opts
+    ]
+
+    Logger.info("[Jackalope] Starting #{__MODULE__} with #{inspect(opts)}")
+    {:ok, pid} = GenServer.start_link(__MODULE__, args)
+    pid
+  end
+
+  @impl GenServer
+  def init(args) do
+    opts = Keyword.fetch!(args, :opts)
+    data_dir = Keyword.get(opts, :data_dir, "/data/jackalope")
+    queue_name = Keyword.get(opts, :queue_name, :items)
+
+    db =
+      case CubDB.start_link(data_dir: data_dir, name: :work_list, auto_compact: true) do
+        {:ok, pid} ->
+          pid
+
+        {:error, {:already_started, pid}} ->
+          pid
+
+        other ->
+          Logger.warn("[Jackalope] Corrupted DB : #{inspect(other)}. Erasing it.")
+          _ = File.rmdir(data_dir)
+          raise "Corrupted work list DB"
+      end
+
+    CubDB.set_auto_file_sync(db, true)
+
+    queue =
+      case CubQ.start_link(db: db, queue: queue_name) do
```

Comment:
I think things could be suffled around and the `CubDB` and `CubQ` `start_links` combined in a helper function that would make testing recovery easier.